packages/actor-optimize-link-traversal-filter-tree-links/lib/ActorOptimizeLinkTraversalFilterTreeLinks.ts

```diff
@@ -0,0 +1,203 @@
+import { BindingsFactory } from '@comunica/bindings-factory';
+import type {
+  IActorOptimizeLinkTraversalArgs,
+  IActionOptimizeLinkTraversal,
+  IActorOptimizeLinkTraversalOutput,
+} from '@comunica/bus-optimize-link-traversal';
+import { ActorOptimizeLinkTraversal } from '@comunica/bus-optimize-link-traversal';
+import { KeysInitQuery } from '@comunica/context-entries';
+import type { IActorTest } from '@comunica/core';
+import type { Bindings } from '@comunica/types';
+import type { IRelation } from '@comunica/types-link-traversal';
+import { DataFactory } from 'rdf-data-factory';
+import type * as RDF from 'rdf-js';
+import { Algebra } from 'sparqlalgebrajs';
+import { AsyncEvaluator } from 'sparqlee';
+
+const DF = new DataFactory<RDF.BaseQuad>();
+
+/**
+ * A comunica Link traversal optimizer that filter link of document
+ * following the [TREE specification](https://treecg.github.io/specification/).
+ * The actor apply the filter of a query into the TREE relation to determine if a
+ * link should be follow or not
+ */
+
+export class ActorOptimizeLinkTraversalFilterTreeLinks extends ActorOptimizeLinkTraversal {
+  public constructor(args: IActorOptimizeLinkTraversalArgs) {
+    super(args);
+  }
+
+  public async test(action: IActionOptimizeLinkTraversal): Promise<IActorTest> {
+    const query: Algebra.Operation = action.context.get(KeysInitQuery.query)!;
+    const relations: IRelation[] = (() => {
+      if (typeof action.treeMetadata !== 'undefined') {
+        return typeof action.treeMetadata.relation !== 'undefined' ? action.treeMetadata.relation : [];
+      }
+      return [];
+    })();
+    const filterExist: boolean = this.doesNodeExist(query, Algebra.types.FILTER);
+    return filterExist && relations.length > 0 ?
+      Promise.resolve(true) :
+      Promise.reject(new Error(
+        'the action must contain TREE relation and the query must contain at least a filter',
+      ));
+  }
+
+  public async run(action: IActionOptimizeLinkTraversal): Promise<IActorOptimizeLinkTraversalOutput> {
+    const filterMap: Map<string, boolean> = new Map();
+    // Extract the filter expression
+    const filterOperation: Algebra.Expression = (() => {
+      const query: Algebra.Operation = action.context.get(KeysInitQuery.query)!;
+      return query.input.expression;
+    })();
+
+    // Extract the bgp of the query
+    const queryBody: RDF.Quad[] = this.findBgp(action.context.get(KeysInitQuery.query)!);
+    if (queryBody.length === 0) {
+      return { filters: filterMap };
+    }
+    // Capture the relation from the input
+    const relations: IRelation[] = (() => {
```

Comment:
I see you apply this pattern several times. Is there a reason why you do this anon function wrapping?

Reply:
This way I can do a double if statement to defined a variable without declaring a temporary one or declaring my variable with a `let`